// @sinclair/typebox-codegen/model/model-to-arktype.js
"use strict";
/*--------------------------------------------------------------------------

@sinclair/typebox-codegen

The MIT License (MIT)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

---------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelToArkType = void 0;
const index_1 = require("../common/index");
const Types = require("@sinclair/typebox");
// --------------------------------------------------------------------------
// ModelToArkType
// --------------------------------------------------------------------------
var ModelToArkType;
(function (ModelToArkType) {
    // ------------------------------------------------------------------------
    // Constraints
    // ------------------------------------------------------------------------
    function Wrap(value) {
        return `'${value}'`;
    }
    function Unwrap(value) {
        if (value.indexOf("'") !== 0)
            return value;
        return value.slice(1, value.length - 1);
    }
    // ------------------------------------------------------------------------
    // Composite
    // ------------------------------------------------------------------------
    function IsTuple(types) {
        return types.some((type) => type.indexOf('[') === 0);
    }
    /**
     * Check if the type need to be wrapped in an api chain (`type(...).some()`)
     * Instead of string.
     */
    function IsApiChain(type) {
        return type.indexOf('type(') === 0;
    }
    function ConstructUnionOrIntersect(types, operator) {
        function Reduce(types) {
            if (types.length === 0)
                return '';
            const [left, ...right] = types;
            return right.length > 0 ? `[${left}, '${operator}', ${Reduce(right)}]` : `${left}`;
        }
        if (IsTuple(types)) {
            return Reduce(types);
        }
        else {
            // For some cases arktype cannot construct types from string literals,
            // i.e `{ [key: string]: any } | undefined` to `'{} | undefined'`.
            // Instead, create api chain `type({}).or('undefined')`.
            const [left, ...right] = types;
            if (IsApiChain(left)) {
                let chainOperator = null;
                if (operator === '&') {
                    chainOperator = 'intersect';
                }
                else if (operator === '|') {
                    chainOperator = 'or';
                }
                return [left, ...right.map((v) => `${chainOperator}(${v})`)].join('.');
            }
            else {
                const mapped = types.map((type) => Unwrap(type)).join(` ${operator} `);
                return Wrap(mapped);
            }
        }
    }
    // ------------------------------------------------------------------------
    // Constraints
    // ------------------------------------------------------------------------
    function ConstrainedSequenceType(type, options) {
        if (IsDefined(options.minimum) && IsDefined(options.maximum)) {
            return Wrap(`${options.minimum}<=${type}<=${options.maximum}`);
        }
        else if (IsDefined(options.minimum)) {
            return Wrap(`${type}>=${options.minimum}`);
        }
        else if (IsDefined(options.maximum)) {
            return Wrap(`${type}<=${options.maximum}`);
        }
        else {
            return Wrap(`${type}`);
        }
    }
    function ConstrainedNumericType(type, options) {
        // prettier-ignore
        const minimum = IsDefined(options.exclusiveMinimum) ? (options.exclusiveMinimum + 1) :
            IsDefined(options.minimum) ? (options.minimum) :
                undefined;
        // prettier-ignore
        const maximum = IsDefined(options.exclusiveMaximum) ? (options.exclusiveMaximum - 1) :
            IsDefined(options.maximum) ? (options.maximum) :
                undefined;
        if (IsDefined(minimum) && IsDefined(maximum)) {
            return Wrap(`${minimum}<=${type}<=${maximum}`);
        }
        else if (IsDefined(minimum)) {
            return Wrap(`${type}>=${minimum}`);
        }
        else if (IsDefined(maximum)) {
            return Wrap(`${type}<=${maximum}`);
        }
        else {
            return Wrap(`${type}`);
        }
    }
    function IsDefined(value) {
        return value !== undefined;
    }
    function Any(schema) {
        return Wrap('any');
    }
    function Array(schema) {
        const type = Visit(schema.items);
        const mapped = `${Unwrap(type)}[]`;
        return ConstrainedSequenceType(mapped, {
            minimum: schema.minItems,
            maximum: schema.maxItems,
        });
    }
    function BigInt(schema) {
        return Wrap('bigint');
    }
    function Boolean(schema) {
        return Wrap('boolean');
    }
    function Constructor(schema) {
        return Wrap('Function');
    }
    function Date(schema) {
        return Wrap('Date');
    }
    function Function(schema) {
        return Wrap('Function');
    }
    function Integer(schema) {
        return ConstrainedNumericType('integer', schema);
    }
    function Intersect(schema) {
        const types = schema.allOf.map((schema) => Collect(schema));
        return ConstructUnionOrIntersect(types, '&');
    }
    function Literal(schema) {
        return typeof schema.const === 'string' ? Wrap(`"${schema.const}"`) : Wrap(schema.const.toString());
    }
    function Never(schema) {
        return Wrap('never');
    }
    function Null(schema) {
        return Wrap('null');
    }
    function String(schema) {
        return ConstrainedSequenceType('string', {
            minimum: schema.minLength,
            maximum: schema.maxLength,
        });
    }
    function Number(schema) {
        return ConstrainedNumericType('number', schema);
    }
    function Object(schema) {
        const properties = globalThis.Object.entries(schema.properties)
            .map(([key, schema]) => {
            const optional = Types.TypeGuard.IsOptional(schema);
            const property1 = index_1.PropertyEncoder.Encode(key);
            const property2 = optional ? `'${property1}?'` : `${property1}`;
            return `${property2}: ${Visit(schema)}`;
        })
            .join(`,`);
        const buffer = [];
        buffer.push(`type({\n${properties}\n})`);
        import_references.add('type');
        return buffer.join(`\n`);
    }
    function Promise(schema) {
        return Wrap('Promise');
    }
    function Record(schema) {
        return Wrap('never'); // not sure how to express
    }
    function Ref(schema) {
        return Wrap(schema.$ref);
    }
    function This(schema) {
        return Wrap(schema.$ref);
    }
    function Tuple(schema) {
        if (schema.items === undefined)
            return `[]`;
        const items = schema.items.map((schema) => Visit(schema)).join(`, `);
        return `[${items}]`;
    }
    function TemplateLiteral(schema) {
        return `/${schema.pattern}/`;
    }
    function UInt8Array(schema) {
        return `['instanceof', Uint8Array]`;
    }
    function Undefined(schema) {
        return Wrap('undefined');
    }
    function Union(schema) {
        const types = schema.anyOf.map((schema) => Collect(schema));
        return ConstructUnionOrIntersect(types, '|');
    }
    function Unknown(schema) {
        return Wrap('unknown');
    }
    function Void(schema) {
        return Wrap('void');
    }
    function UnsupportedType(schema) {
        return Wrap('never');
    }
    function Visit(schema) {
        if (schema.$id !== undefined)
            reference_map.set(schema.$id, schema);
        if (schema.$id !== undefined && emitted_types.has(schema.$id))
            return `'${schema.$id}'`;
        if (Types.TypeGuard.IsAny(schema))
            return Any(schema);
        if (Types.TypeGuard.IsArray(schema))
            return Array(schema);
        if (Types.TypeGuard.IsBigInt(schema))
            return BigInt(schema);
        if (Types.TypeGuard.IsBoolean(schema))
            return Boolean(schema);
        if (Types.TypeGuard.IsConstructor(schema))
            return Constructor(schema);
        if (Types.TypeGuard.IsDate(schema))
            return Date(schema);
        if (Types.TypeGuard.IsFunction(schema))
            return Function(schema);
        if (Types.TypeGuard.IsInteger(schema))
            return Integer(schema);
        if (Types.TypeGuard.IsIntersect(schema))
            return Intersect(schema);
        if (Types.TypeGuard.IsLiteral(schema))
            return Literal(schema);
        if (Types.TypeGuard.IsNever(schema))
            return Never(schema);
        if (Types.TypeGuard.IsNull(schema))
            return Null(schema);
        if (Types.TypeGuard.IsNumber(schema))
            return Number(schema);
        if (Types.TypeGuard.IsObject(schema))
            return Object(schema);
        if (Types.TypeGuard.IsPromise(schema))
            return Promise(schema);
        if (Types.TypeGuard.IsRecord(schema))
            return Record(schema);
        if (Types.TypeGuard.IsRef(schema))
            return Ref(schema);
        if (Types.TypeGuard.IsString(schema))
            return String(schema);
        if (Types.TypeGuard.IsTemplateLiteral(schema))
            return TemplateLiteral(schema);
        if (Types.TypeGuard.IsThis(schema))
            return This(schema);
        if (Types.TypeGuard.IsTuple(schema))
            return Tuple(schema);
        if (Types.TypeGuard.IsUint8Array(schema))
            return UInt8Array(schema);
        if (Types.TypeGuard.IsUndefined(schema))
            return Undefined(schema);
        if (Types.TypeGuard.IsUnion(schema))
            return Union(schema);
        if (Types.TypeGuard.IsUnknown(schema))
            return Unknown(schema);
        if (Types.TypeGuard.IsVoid(schema))
            return Void(schema);
        return UnsupportedType(schema);
    }
    function Collect(schema) {
        return [...Visit(schema)].join(``);
    }
    function GenerateType(schema, references) {
        const buffer = [];
        for (const reference of references) {
            if (reference.$id === undefined)
                return UnsupportedType(schema);
            reference_map.set(reference.$id, reference);
        }
        const type = Collect(schema);
        buffer.push(`${schema.$id || `T`}: ${type}`);
        if (schema.$id)
            emitted_types.add(schema.$id);
        return buffer.join('\n');
    }
    const reference_map = new Map();
    const emitted_types = new Set();
    const import_references = new Set();
    function Generate(model) {
        reference_map.clear();
        emitted_types.clear();
        const buffer = [];
        buffer.push('export const types = scope({');
        for (const type of model.types.filter((type) => Types.TypeGuard.IsSchema(type))) {
            buffer.push(`${GenerateType(type, model.types)},`);
        }
        buffer.push('}).export()');
        buffer.push('\n');
        for (const type of model.types.filter((type) => Types.TypeGuard.IsSchema(type))) {
            buffer.push(`export type ${type.$id} = typeof ${type.$id}.infer`);
            buffer.push(`export const ${type.$id} = types.${type.$id}`);
        }
        const imports = ['scope', ...import_references].join(',');
        buffer.unshift(`import { ${imports} } from 'arktype'`, '');
        import_references.clear();
        return index_1.Formatter.Format(buffer.join('\n'));
    }
    ModelToArkType.Generate = Generate;
})(ModelToArkType || (exports.ModelToArkType = ModelToArkType = {}));
