// @sinclair/typebox-codegen/model/model-to-valibot.js
"use strict";
/*--------------------------------------------------------------------------

@sinclair/typebox-codegen

The MIT License (MIT)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

---------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelToValibot = void 0;
const index_1 = require("../common/index");
const model_to_typescript_1 = require("./model-to-typescript");
const Types = require("@sinclair/typebox");
// --------------------------------------------------------------------------
// ModelToValibot
// --------------------------------------------------------------------------
// prettier-ignore
var ModelToValibot;
(function (ModelToValibot) {
    function IsDefined(value) {
        return value !== undefined;
    }
    function Type(type, parameter, constraints) {
        if (constraints.length > 0) {
            if (typeof parameter === 'string') {
                return `v.pipe(${type}(${parameter}), ${constraints.join(', ')})`;
            }
            else {
                return `v.pipe(${type}(), ${constraints.join(', ')})`;
            }
        }
        else {
            if (typeof parameter === 'string') {
                return `${type}(${parameter})`;
            }
            else {
                return `${type}()`;
            }
        }
    }
    function Any(schema) {
        return Type(`v.any`, null, []);
    }
    function Array(schema) {
        const items = Visit(schema.items);
        const constraints = [];
        if (IsDefined(schema.minItems))
            constraints.push(`v.minLength(${schema.minItems})`);
        if (IsDefined(schema.maxItems))
            constraints.push(`v.maxLength(${schema.maxItems})`);
        return Type(`v.array`, items, constraints);
    }
    function BigInt(schema) {
        return Type(`v.bigint`, null, []);
    }
    function Boolean(schema) {
        return Type(`v.boolean`, null, []);
    }
    function Date(schema) {
        return Type(`v.date`, null, []);
    }
    function Constructor(schema) {
        return UnsupportedType(schema);
    }
    function Function(schema) {
        return UnsupportedType(schema);
    }
    function Integer(schema) {
        return Type(`v.number`, null, [`v.integer()`]);
    }
    function Intersect(schema) {
        const inner = schema.allOf.map((inner) => Visit(inner));
        return Type(`v.intersect`, `[${inner.join(', ')}]`, []);
    }
    function Literal(schema) {
        return typeof schema.const === `string`
            ? Type(`v.literal`, `'${schema.const}'`, [])
            : Type(`v.literal`, `${schema.const}`, []);
    }
    function Never(schema) {
        return Type(`v.never`, null, []);
    }
    function Null(schema) {
        return Type(`v.null`, null, []);
    }
    function String(schema) {
        const constraints = [];
        if (IsDefined(schema.maxLength))
            constraints.push(`v.maxLength(${schema.maxLength})`);
        if (IsDefined(schema.minLength))
            constraints.push(`v.minLength(${schema.minLength})`);
        return Type(`v.string`, null, constraints);
    }
    function Number(schema) {
        const constraints = [];
        if (IsDefined(schema.minimum))
            constraints.push(`v.minValue(${schema.minimum})`);
        if (IsDefined(schema.maximum))
            constraints.push(`v.maxValue(${schema.maximum})`);
        if (IsDefined(schema.exclusiveMinimum))
            constraints.push(`v.minValue(${schema.exclusiveMinimum + 1})`);
        if (IsDefined(schema.exclusiveMaximum))
            constraints.push(`v.maxValue(${schema.exclusiveMaximum - 1})`);
        return Type('v.number', null, constraints);
    }
    function Object(schema) {
        const properties = globalThis.Object.entries(schema.properties).map(([key, value]) => {
            const optional = Types.TypeGuard.IsOptional(value);
            const property = index_1.PropertyEncoder.Encode(key);
            return optional
                ? settings.exactOptionalPropertyTypes
                    ? `${property}: v.exactOptional(${Visit(value)})`
                    : `${property}: v.optional(${Visit(value)})`
                : `${property}: ${Visit(value)}`;
        }).join(`,`);
        const constraints = [];
        return Type(`v.object`, `{\n${properties}\n}`, constraints);
    }
    function Promise(schema) {
        return UnsupportedType(schema);
    }
    function Record(schema) {
        for (const [key, value] of globalThis.Object.entries(schema.patternProperties)) {
            const type = Visit(value);
            if (key === `^(0|[1-9][0-9]*)$`) {
                return Type('v.record', `v.number(), ${type}`, []);
            }
            else {
                return Type(`v.record`, `v.string(), ${type}`, []);
            }
        }
        throw Error(`Unreachable`);
    }
    function Ref(schema) {
        if (!reference_map.has(schema.$ref))
            return UnsupportedType(schema);
        return schema.$ref;
    }
    function This(schema) {
        return UnsupportedType(schema);
    }
    function Tuple(schema) {
        if (schema.items === undefined)
            return `[]`;
        const items = schema.items.map((schema) => Visit(schema)).join(`, `);
        return Type(`v.tuple`, `[${items}]`, []);
    }
    function TemplateLiteral(schema) {
        const constaint = Type(`v.regex`, `/${schema.pattern}/`, []);
        return Type(`v.string`, null, [constaint]);
    }
    function UInt8Array(schema) {
        return UnsupportedType(schema);
    }
    function Undefined(schema) {
        return Type(`v.undefined`, null, []);
    }
    function Union(schema) {
        const inner = schema.anyOf.map((schema) => Visit(schema)).join(`, `);
        return Type(`v.union`, `[${inner}]`, []);
    }
    function Unknown(schema) {
        return Type(`v.unknown`, null, []);
    }
    function Void(schema) {
        return Type(`v.void`, null, []);
    }
    function UnsupportedType(schema) {
        return `v.any(/* unsupported */)`;
    }
    function Visit(schema) {
        if (schema.$id !== undefined)
            reference_map.set(schema.$id, schema);
        if (schema.$id !== undefined && emitted_set.has(schema.$id))
            return schema.$id;
        if (Types.TypeGuard.IsAny(schema))
            return Any(schema);
        if (Types.TypeGuard.IsArray(schema))
            return Array(schema);
        if (Types.TypeGuard.IsBigInt(schema))
            return BigInt(schema);
        if (Types.TypeGuard.IsBoolean(schema))
            return Boolean(schema);
        if (Types.TypeGuard.IsDate(schema))
            return Date(schema);
        if (Types.TypeGuard.IsConstructor(schema))
            return Constructor(schema);
        if (Types.TypeGuard.IsFunction(schema))
            return Function(schema);
        if (Types.TypeGuard.IsInteger(schema))
            return Integer(schema);
        if (Types.TypeGuard.IsIntersect(schema))
            return Intersect(schema);
        if (Types.TypeGuard.IsLiteral(schema))
            return Literal(schema);
        if (Types.TypeGuard.IsNever(schema))
            return Never(schema);
        if (Types.TypeGuard.IsNull(schema))
            return Null(schema);
        if (Types.TypeGuard.IsNumber(schema))
            return Number(schema);
        if (Types.TypeGuard.IsObject(schema))
            return Object(schema);
        if (Types.TypeGuard.IsPromise(schema))
            return Promise(schema);
        if (Types.TypeGuard.IsRecord(schema))
            return Record(schema);
        if (Types.TypeGuard.IsRef(schema))
            return Ref(schema);
        if (Types.TypeGuard.IsString(schema))
            return String(schema);
        if (Types.TypeGuard.IsTemplateLiteral(schema))
            return TemplateLiteral(schema);
        if (Types.TypeGuard.IsThis(schema))
            return This(schema);
        if (Types.TypeGuard.IsTuple(schema))
            return Tuple(schema);
        if (Types.TypeGuard.IsUint8Array(schema))
            return UInt8Array(schema);
        if (Types.TypeGuard.IsUndefined(schema))
            return Undefined(schema);
        if (Types.TypeGuard.IsUnion(schema))
            return Union(schema);
        if (Types.TypeGuard.IsUnknown(schema))
            return Unknown(schema);
        if (Types.TypeGuard.IsVoid(schema))
            return Void(schema);
        return UnsupportedType(schema);
    }
    function Collect(schema) {
        return [...Visit(schema)].join(``);
    }
    function GenerateType(model, schema, references) {
        const output = [];
        for (const reference of references) {
            if (reference.$id === undefined)
                return UnsupportedType(schema);
            reference_map.set(reference.$id, reference);
        }
        const type = Collect(schema);
        if (recursive_set.has(schema.$id)) {
            output.push(`export ${model_to_typescript_1.ModelToTypeScript.GenerateType(model, schema.$id)}`);
            output.push(`export const ${schema.$id || `T`}: v.InferOutput<${schema.$id}> = v.lazy(() => ${index_1.Formatter.Format(type)})`);
        }
        else {
            output.push(`export type ${schema.$id} = v.InferOutput<typeof ${schema.$id}>`);
            output.push(`export const ${schema.$id || `T`} = ${index_1.Formatter.Format(type)}`);
        }
        if (schema.$id)
            emitted_set.add(schema.$id);
        return output.join('\n');
    }
    const reference_map = new Map();
    const recursive_set = new Set();
    const emitted_set = new Set();
    const settings = {
        exactOptionalPropertyTypes: false
    };
    function Generate(model, options = {
        exactOptionalPropertyTypes: false
    }) {
        settings.exactOptionalPropertyTypes = options.exactOptionalPropertyTypes;
        reference_map.clear();
        recursive_set.clear();
        emitted_set.clear();
        const buffer = [`import * as v from 'valibot'`, ''];
        for (const type of model.types.filter((type) => Types.TypeGuard.IsSchema(type))) {
            buffer.push(GenerateType(model, type, model.types));
        }
        return index_1.Formatter.Format(buffer.join('\n'));
    }
    ModelToValibot.Generate = Generate;
})(ModelToValibot || (exports.ModelToValibot = ModelToValibot = {}));
