// @sinclair/typebox-codegen/model/model-to-zod.js
"use strict";
/*--------------------------------------------------------------------------

@sinclair/typebox-codegen

The MIT License (MIT)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

---------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelToZod = void 0;
const index_1 = require("../common/index");
const model_to_typescript_1 = require("./model-to-typescript");
const Types = require("@sinclair/typebox");
// --------------------------------------------------------------------------
// ModelToZod
// --------------------------------------------------------------------------
var ModelToZod;
(function (ModelToZod) {
    function IsDefined(value) {
        return value !== undefined;
    }
    function Type(schema, type) {
        return schema.default === undefined ? type : `${type}.default(${JSON.stringify(schema.default)})`;
    }
    function Any(schema) {
        return Type(schema, `z.any()`);
    }
    function Array(schema) {
        const items = Visit(schema.items);
        const buffer = [];
        buffer.push(`z.array(${items})`);
        if (IsDefined(schema.minItems))
            buffer.push(`.min(${schema.minItems})`);
        if (IsDefined(schema.maxItems))
            buffer.push(`.max(${schema.maxItems})`);
        return Type(schema, buffer.join(``));
    }
    function BigInt(schema) {
        return Type(schema, `z.bigint()`);
    }
    function Boolean(schema) {
        return Type(schema, `z.boolean()`);
    }
    function Date(schema) {
        return Type(schema, `z.date()`);
    }
    function Constructor(schema) {
        return UnsupportedType(schema);
    }
    function Function(schema) {
        const params = schema.parameters.map((param) => Visit(param)).join(`, `);
        const returns = Visit(schema.returns);
        return Type(schema, `z.function().args(${params}).returns(${returns})`);
    }
    function Integer(schema) {
        const buffer = [];
        buffer.push(`z.number().int()`);
        if (IsDefined(schema.minimum))
            buffer.push(`.min(${schema.minimum})`);
        if (IsDefined(schema.maximum))
            buffer.push(`.max(${schema.maximum})`);
        if (IsDefined(schema.exclusiveMinimum))
            buffer.push(`.min(${schema.exclusiveMinimum + 1})`);
        if (IsDefined(schema.exclusiveMaximum))
            buffer.push(`.max(${schema.exclusiveMaximum - 1})`);
        if (IsDefined(schema.multipleOf))
            buffer.push(`.multipleOf(${schema.multipleOf})`);
        return Type(schema, buffer.join(``));
    }
    function Intersect(schema) {
        function reduce(rest) {
            if (rest.length === 0)
                return `z.never()`;
            if (rest.length === 1)
                return Visit(rest[0]);
            const [left, right] = [rest[0], rest.slice(1)];
            return Type(schema, `z.intersection(${Visit(left)}, ${reduce(right)})`);
        }
        return reduce(schema.allOf);
    }
    function Literal(schema) {
        return typeof schema.const === `string` ? Type(schema, `z.literal('${schema.const}')`) : Type(schema, `z.literal(${schema.const})`);
    }
    function Never(schema) {
        return Type(schema, `z.never()`);
    }
    function Null(schema) {
        return Type(schema, `z.null()`);
    }
    function String(schema) {
        const buffer = [];
        buffer.push(`z.string()`);
        if (IsDefined(schema.maxLength))
            buffer.push(`.max(${schema.maxLength})`);
        if (IsDefined(schema.minLength))
            buffer.push(`.min(${schema.minLength})`);
        return Type(schema, buffer.join(``));
    }
    function Number(schema) {
        const buffer = [];
        buffer.push(`z.number()`);
        if (IsDefined(schema.minimum))
            buffer.push(`.min(${schema.minimum})`);
        if (IsDefined(schema.maximum))
            buffer.push(`.max(${schema.maximum})`);
        if (IsDefined(schema.exclusiveMinimum))
            buffer.push(`.min(${schema.exclusiveMinimum + 1})`);
        if (IsDefined(schema.exclusiveMaximum))
            buffer.push(`.max(${schema.exclusiveMaximum - 1})`);
        if (IsDefined(schema.multipleOf))
            buffer.push(`.multipleOf(${schema.multipleOf})`);
        return Type(schema, buffer.join(``));
    }
    function Object(schema) {
        // prettier-ignore
        const properties = globalThis.Object.entries(schema.properties).map(([key, value]) => {
            const optional = Types.TypeGuard.IsOptional(value);
            const readonly = Types.TypeGuard.IsReadonly(value);
            const property = index_1.PropertyEncoder.Encode(key);
            // prettier-ignore
            return (readonly && optional ? `${property}: ${Visit(value)}.readonly().optional()` :
                readonly ? `${property}: ${Visit(value)}.readonly()` :
                    optional ? `${property}: ${Visit(value)}.optional()` :
                        `${property}: ${Visit(value)}`);
        }).join(`,`);
        const buffer = [];
        buffer.push(`z.object({\n${properties}\n})`);
        if (schema.additionalProperties === false)
            buffer.push(`.strict()`);
        return Type(schema, buffer.join(``));
    }
    function Promise(schema) {
        const item = Visit(schema.item);
        return Type(schema, `${item}.promise()`);
    }
    function Record(schema) {
        for (const [key, value] of globalThis.Object.entries(schema.patternProperties)) {
            const type = Visit(value);
            if (key === `^(0|[1-9][0-9]*)$`) {
                return `z.record(z.number(), ${type})`;
            }
            else {
                return `z.record(${type})`;
            }
        }
        throw Error(`Unreachable`);
    }
    function Ref(schema) {
        if (!reference_map.has(schema.$ref))
            return UnsupportedType(schema); // throw new ModelToZodNonReferentialType(schema.$ref!)
        return schema.$ref;
    }
    function This(schema) {
        if (!reference_map.has(schema.$ref))
            return UnsupportedType(schema); //throw new ModelToZodNonReferentialType(schema.$ref!)
        recursive_set.add(schema.$ref);
        return schema.$ref;
    }
    function Tuple(schema) {
        if (schema.items === undefined)
            return `[]`;
        const items = schema.items.map((schema) => Visit(schema)).join(`, `);
        return Type(schema, `z.tuple([${items}])`);
    }
    function TemplateLiteral(schema) {
        return Type(schema, `z.string().regex(/${schema.pattern}/)`);
    }
    function UInt8Array(schema) {
        return Type(schema, `z.instanceof(Uint8Array)`);
    }
    function Undefined(schema) {
        return Type(schema, `z.undefined()`);
    }
    function Union(schema) {
        return Type(schema, `z.union([${schema.anyOf.map((schema) => Visit(schema)).join(`, `)}])`);
    }
    function Unknown(schema) {
        return Type(schema, `z.unknown()`);
    }
    function Void(schema) {
        return Type(schema, `z.void()`);
    }
    function UnsupportedType(schema) {
        return `${Type(schema, `z.any( /* unsupported */)`)}`;
    }
    function Visit(schema) {
        if (schema.$id !== undefined)
            reference_map.set(schema.$id, schema);
        if (schema.$id !== undefined && emitted_set.has(schema.$id))
            return schema.$id;
        if (Types.TypeGuard.IsAny(schema))
            return Any(schema);
        if (Types.TypeGuard.IsArray(schema))
            return Array(schema);
        if (Types.TypeGuard.IsBigInt(schema))
            return BigInt(schema);
        if (Types.TypeGuard.IsBoolean(schema))
            return Boolean(schema);
        if (Types.TypeGuard.IsDate(schema))
            return Date(schema);
        if (Types.TypeGuard.IsConstructor(schema))
            return Constructor(schema);
        if (Types.TypeGuard.IsFunction(schema))
            return Function(schema);
        if (Types.TypeGuard.IsInteger(schema))
            return Integer(schema);
        if (Types.TypeGuard.IsIntersect(schema))
            return Intersect(schema);
        if (Types.TypeGuard.IsLiteral(schema))
            return Literal(schema);
        if (Types.TypeGuard.IsNever(schema))
            return Never(schema);
        if (Types.TypeGuard.IsNull(schema))
            return Null(schema);
        if (Types.TypeGuard.IsNumber(schema))
            return Number(schema);
        if (Types.TypeGuard.IsObject(schema))
            return Object(schema);
        if (Types.TypeGuard.IsPromise(schema))
            return Promise(schema);
        if (Types.TypeGuard.IsRecord(schema))
            return Record(schema);
        if (Types.TypeGuard.IsRef(schema))
            return Ref(schema);
        if (Types.TypeGuard.IsString(schema))
            return String(schema);
        if (Types.TypeGuard.IsTemplateLiteral(schema))
            return TemplateLiteral(schema);
        if (Types.TypeGuard.IsThis(schema))
            return This(schema);
        if (Types.TypeGuard.IsTuple(schema))
            return Tuple(schema);
        if (Types.TypeGuard.IsUint8Array(schema))
            return UInt8Array(schema);
        if (Types.TypeGuard.IsUndefined(schema))
            return Undefined(schema);
        if (Types.TypeGuard.IsUnion(schema))
            return Union(schema);
        if (Types.TypeGuard.IsUnknown(schema))
            return Unknown(schema);
        if (Types.TypeGuard.IsVoid(schema))
            return Void(schema);
        return UnsupportedType(schema);
    }
    function Collect(schema) {
        return [...Visit(schema)].join(``);
    }
    function GenerateType(model, schema, references) {
        const output = [];
        for (const reference of references) {
            if (reference.$id === undefined)
                return UnsupportedType(schema);
            reference_map.set(reference.$id, reference);
        }
        const type = Collect(schema);
        if (recursive_set.has(schema.$id)) {
            output.push(`export ${model_to_typescript_1.ModelToTypeScript.GenerateType(model, schema.$id)}`);
            output.push(`export const ${schema.$id || `T`}: z.ZodType<${schema.$id}> = z.lazy(() => ${index_1.Formatter.Format(type)})`);
        }
        else {
            output.push(`export type ${schema.$id} = z.infer<typeof ${schema.$id}>`);
            output.push(`export const ${schema.$id || `T`} = ${index_1.Formatter.Format(type)}`);
        }
        if (schema.$id)
            emitted_set.add(schema.$id);
        return output.join('\n');
    }
    const reference_map = new Map();
    const recursive_set = new Set();
    const emitted_set = new Set();
    function Generate(model) {
        reference_map.clear();
        recursive_set.clear();
        emitted_set.clear();
        const buffer = [`import { z } from 'zod'`, ''];
        for (const type of model.types.filter((type) => Types.TypeGuard.IsSchema(type))) {
            buffer.push(GenerateType(model, type, model.types));
        }
        return index_1.Formatter.Format(buffer.join('\n'));
    }
    ModelToZod.Generate = Generate;
})(ModelToZod || (exports.ModelToZod = ModelToZod = {}));
