// @sinclair/typebox-codegen/model/model-to-json-schema.js
"use strict";
/*--------------------------------------------------------------------------

@sinclair/typebox-codegen

The MIT License (MIT)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

---------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelToJsonSchema = void 0;
const index_1 = require("../common/index");
const Types = require("@sinclair/typebox");
// --------------------------------------------------------------------------
// ModelToJsonSchema
// --------------------------------------------------------------------------
var ModelToJsonSchema;
(function (ModelToJsonSchema) {
    function Any(schema) {
        return schema;
    }
    function Array(schema) {
        return schema;
    }
    function BigInt(schema) {
        return schema;
    }
    function Boolean(schema) {
        return schema;
    }
    function Date(schema) {
        return schema;
    }
    function Constructor(schema) {
        return schema;
    }
    function Function(schema) {
        const parameters = schema.parameters.map((schema) => Visit(schema));
        const returns = Visit(schema.returns);
        return { ...schema, parameters, returns };
    }
    function Integer(schema) {
        return schema;
    }
    function Intersect(schema) {
        const allOf = schema.allOf.map((schema) => Visit(schema));
        return { ...schema, allOf };
    }
    function Literal(schema) {
        return schema;
    }
    function Never(schema) {
        return schema;
    }
    function Null(schema) {
        return schema;
    }
    function String(schema) {
        return schema;
    }
    function Number(schema) {
        return schema;
    }
    function Object(schema) {
        const properties = globalThis.Object.keys(schema.properties).reduce((acc, key) => {
            return { ...acc, [key]: Visit(schema.properties[key]) };
        }, {});
        return { ...schema, properties };
    }
    function Promise(schema) {
        const item = Visit(schema.item);
        return { ...schema, item };
    }
    function Record(schema) {
        const patternProperties = globalThis.Object.keys(schema.patternProperties).reduce((acc, key) => {
            return { ...acc, [key]: Visit(schema.patternProperties[key]) };
        }, {});
        return { ...schema, patternProperties };
    }
    function Ref(schema) {
        return schema;
    }
    function This(schema) {
        return schema;
    }
    function Tuple(schema) {
        if (schema.items === undefined)
            return schema;
        const items = schema.items.map((schema) => Visit(schema));
        return { ...schema, items };
    }
    function TemplateLiteral(schema) {
        return schema;
    }
    function UInt8Array(schema) {
        return schema;
    }
    function Undefined(schema) {
        return schema;
    }
    function Union(schema) {
        const anyOf = schema.anyOf.map((schema) => Visit(schema));
        return { ...schema, anyOf };
    }
    function Unknown(schema) {
        return schema;
    }
    function Void(schema) {
        return schema;
    }
    function UnsupportedType(schema) {
        return schema;
    }
    function Visit(schema) {
        if (Types.TypeGuard.IsAny(schema))
            return Any(schema);
        if (Types.TypeGuard.IsArray(schema))
            return Array(schema);
        if (Types.TypeGuard.IsBigInt(schema))
            return BigInt(schema);
        if (Types.TypeGuard.IsBoolean(schema))
            return Boolean(schema);
        if (Types.TypeGuard.IsDate(schema))
            return Date(schema);
        if (Types.TypeGuard.IsConstructor(schema))
            return Constructor(schema);
        if (Types.TypeGuard.IsFunction(schema))
            return Function(schema);
        if (Types.TypeGuard.IsInteger(schema))
            return Integer(schema);
        if (Types.TypeGuard.IsIntersect(schema))
            return Intersect(schema);
        if (Types.TypeGuard.IsLiteral(schema))
            return Literal(schema);
        if (Types.TypeGuard.IsNever(schema))
            return Never(schema);
        if (Types.TypeGuard.IsNull(schema))
            return Null(schema);
        if (Types.TypeGuard.IsNumber(schema))
            return Number(schema);
        if (Types.TypeGuard.IsObject(schema))
            return Object(schema);
        if (Types.TypeGuard.IsPromise(schema))
            return Promise(schema);
        if (Types.TypeGuard.IsRecord(schema))
            return Record(schema);
        if (Types.TypeGuard.IsRef(schema))
            return Ref(schema);
        if (Types.TypeGuard.IsString(schema))
            return String(schema);
        if (Types.TypeGuard.IsTemplateLiteral(schema))
            return TemplateLiteral(schema);
        if (Types.TypeGuard.IsThis(schema))
            return This(schema);
        if (Types.TypeGuard.IsTuple(schema))
            return Tuple(schema);
        if (Types.TypeGuard.IsUint8Array(schema))
            return UInt8Array(schema);
        if (Types.TypeGuard.IsUndefined(schema))
            return Undefined(schema);
        if (Types.TypeGuard.IsUnion(schema))
            return Union(schema);
        if (Types.TypeGuard.IsUnknown(schema))
            return Unknown(schema);
        if (Types.TypeGuard.IsVoid(schema))
            return Void(schema);
        return UnsupportedType(schema);
    }
    function Generate(model) {
        const buffer = [];
        for (const type of model.types.filter((type) => Types.TypeGuard.IsSchema(type))) {
            const schema = Visit(type);
            const encode = JSON.stringify(schema, null, 2);
            buffer.push(`export const ${type.$id} = ${encode}`);
        }
        return index_1.Formatter.Format(buffer.join('\n'));
    }
    ModelToJsonSchema.Generate = Generate;
})(ModelToJsonSchema || (exports.ModelToJsonSchema = ModelToJsonSchema = {}));
